aiohttp/_multidict.pyx

```diff
@@ -0,0 +1,500 @@
+            if isinstance(other, MultiDict):
+                typed_other = other
+                return typed_self._items != typed_other._items
+            for item in typed_self._items:
+                nv = other.get(item[0], _marker)
+                if item[1] == nv:
+                    return True
+            return False
+        else:
+            return NotImplemented
+
+    def __repr__(self):
+        return '<{}>\n{}'.format(
+            self.__class__.__name__, pprint.pformat(self._items))
+
+
+abc.Mapping.register(MultiDict)
+
+
+cdef class CaseInsensitiveMultiDict(MultiDict):
+    """Case insensitive multi dict."""
+
+    @classmethod
+    def _from_uppercase_multidict(cls, MultiDict dct):
+        # NB: doesn't check for uppercase keys!
+        cdef CaseInsensitiveMultiDict ret
+        ret = cls.__new__(cls)
+        ret._items = dct._items
+        return ret
+
+    cdef _upper(self, s):
+        if type(s) is cistr:
+            return s
+        return s.upper()
+
+    cdef _add(self, tuple item):
+        self._items.append((self._upper(item[0]), item[1]))
+
+    def getall(self, key, default=_marker):
+        return self._getall(self._upper(key), default)
+
+    def getone(self, key, default=_marker):
+        return self._getone(self._upper(key), default)
+
+    def get(self, key, default=None):
+        return self._get(self._upper(key), default)
+
+    def __getitem__(self, key):
+        return self._getitem(self._upper(key))
+
+    def __contains__(self, key):
+        return self._contains(self._upper(key))
+
+
+abc.Mapping.register(CaseInsensitiveMultiDict)
+
+
+cdef class MutableMultiDict(MultiDict):
+    """An ordered dictionary that can have multiple values for each key."""
+
+    def add(self, key, value):
+        """
+        Add the key and value, not overwriting any previous value.
+        """
+        self._add((key, value))
+
+    def extend(self, *args, **kwargs):
+        """Extends current MutableMultiDict with more values.
+
+        This method must be used instead of update.
+        """
+        self._extend(args, kwargs, "extend")
+
+    def clear(self):
+        """Remove all items from MutableMultiDict"""
+        self._items = []
+
+    # MutableMapping interface #
+
+    def __setitem__(self, key, value):
+        self._delitem(key, False)
+        self._add((key, value))
+
+    def __delitem__(self, key):
+        self._delitem(key, True)
+
+    def setdefault(self, key, default=None):
+        for k, v in self._items:
+            if k == key:
+                return v
+        self._add((key, default))
+        return default
+
+    def pop(self, key, default=None):
+        """Method not allowed."""
+        raise NotImplementedError
+
+    def popitem(self):
+        """Method not allowed."""
+        raise NotImplementedError
+
+    def update(self, *args, **kw):
+        """Method not allowed."""
+        raise NotImplementedError("Use extend method instead")
```

Comment:
Why not make update as proxy to extend? Current implementation breaks possibility to use plain old dicts for headers because MultiDict form one side acts as dict, but on this point it does not.